subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c

```diff
@@ -161,13 +161,43 @@ gst_msdkdec_free_unlocked_msdk_surfaces (GstMsdkDec * thiz)
   for (l = thiz->locked_msdk_surfaces; l;) {
     GList *next = l->next;
     surface = l->data;
-    if (surface->surface->Data.Locked == 0) {
+    if (surface->surface->Data.Locked == 0 &&
+        GST_MINI_OBJECT_REFCOUNT (surface->buf) == 1) {
       free_surface (surface);
       thiz->locked_msdk_surfaces =
           g_list_delete_link (thiz->locked_msdk_surfaces, l);
     }
     l = next;
   }
+  /* We need to check if all surfaces are in used */
+  if (g_list_length (thiz->locked_msdk_surfaces) ==
+      thiz->alloc_resp.NumFrameActual)
+    return FALSE;
+  else
+    return TRUE;
+}
+
+static gboolean
+check_surface_available (GstMsdkDec * thiz)
```

Comment:
It is very similar with the previous gst_msdkdec_free_unlocked_msdk_surfaces(), can you merge them together?

Reply:
The func gst_msdkdec_free_unlocked_msdk_surfaces() aims to free all un-used msdk surfaces back to the pool before we acquire a buffer from the pool while check_surface_available() just needs to find one available surface. So, here I  merge this two and add a flag (check_avail_surface) to decide if we need to find all (flag is TRUE) or just one surface (flag is FALSE).